cpp/rendler_v1_executor.cpp

```diff
@@ -0,0 +1,232 @@
+          launchTask(task);
+        }
+        break;
+      }
+
+      case Event::KILL: {
+        break;
+      }
+
+      case Event::ACKNOWLEDGED: {
+        // Remove the corresponding update.
+        updates.erase(UUID::fromBytes(event.acknowledged().uuid()).get());
+        // Remove the corresponding task.
+        tasks.erase(event.acknowledged().task_id());
+        break;
+      }
+
+      case Event::MESSAGE: {
+        break;
+      }
+
+      case Event::SHUTDOWN: {
+        break;
+      }
+
+      case Event::ERROR: {
+        break;
+      }
+
+      case Event::UNKNOWN: {
+        LOG(WARNING) << "Received an UNKNOWN event and ignored";
+        break;
+      }
+    }
+  }
+}
+
+
+void RendlerV1Executor::sendStatusUpdate(const TaskInfo& task,
+                                         const TaskState& state)
+{
+  UUID uuid = UUID::random();
+
+  TaskStatus status;
+  status.mutable_task_id()->CopyFrom(task.task_id());
+  status.mutable_executor_id()->CopyFrom(executorId);
+  status.set_state(state);
+  status.set_source(TaskStatus::SOURCE_EXECUTOR);
+  status.set_uuid(uuid.toBytes());
+
+  Call call;
+  call.mutable_framework_id()->CopyFrom(frameworkId);
+  call.mutable_executor_id()->CopyFrom(executorId);
+  call.set_type(Call::UPDATE);
+
+  Call::Update* update = call.mutable_update();
+  update->mutable_status()->CopyFrom(status);
+
+  // Capture the status update so we can repeat if there is no ack.
+  updates[uuid] = call.update();
+
+  mesos->send(call);
+}
+
+
+void RendlerV1Executor::sendFrameworkMessage(const string& data)
+{
+  Call call;
+  call.mutable_framework_id()->CopyFrom(frameworkId);
+  call.mutable_executor_id()->CopyFrom(executorId);
+  call.set_type(Call::MESSAGE);
+
+  Call::Message* message = call.mutable_message();
+  message->set_data(data);
+
+  mesos->send(call);
+}
+
+
+void RendlerV1Executor::initialize()
+{
+  // We initialize the library here to ensure that callbacks are only invoked
+  // after the process has spawned.
+  mesos.reset(new Mesos(
+      mesos::ContentType::PROTOBUF,
+      process::defer(self(), &Self::connected),
+      process::defer(self(), &Self::disconnected),
+      process::defer(self(), &Self::received, lambda::_1)));
+}
+
+
+void RendlerV1Executor::doReliableRegistration()
+{
+  if (state == SUBSCRIBED || state == DISCONNECTED) {
+    return;
+  }
+
+  Call call;
+  call.mutable_framework_id()->CopyFrom(frameworkId);
+  call.mutable_executor_id()->CopyFrom(executorId);
+
+  call.set_type(Call::SUBSCRIBE);
+
+  Call::Subscribe* subscribe = call.mutable_subscribe();
+
+  // Send all unacknowledged updates.
+  foreachvalue (const Call::Update& update, updates) {
+    subscribe->add_unacknowledged_updates()->MergeFrom(update);
+  }
+
+  // Send all unacknowledged tasks.
+  foreachvalue (const TaskInfo& task, tasks) {
+    subscribe->add_unacknowledged_tasks()->MergeFrom(task);
+  }
+
+  mesos->send(call);
+
+  // Re-registrate after 1 second if not subscribed then.
```

Comment:
s/Re-registrate/Re-register
Also
"Re-register after one second if not subscribed by then."